### ISSUE TITLE
Company-lsp failing to auto-begin in typescript-mode and js2-mode, disabling all completions

### ISSUE THREAD

**Issue by cmatzenbach** [emacs-lsp/lsp-mode#513]
**Describe the bug**
When loading a .ts or .js file using the latest versions of lsp-mode and company-lsp, company fails to initialize correctly.  Upon the first instance of `company-complete`, I receive an error message in the minibuffer saying `Company: backend company-lsp error "Wrong type argument: lsp--workspace, nil" with args (prefix)`.  Upon inspection of the messages buffer, I see the same error, but with this additional line above it: `Company: An error occurred in auto-begin`.  Those are the only company-related messages in the `*Messages*` buffer, and that one error is all I see in the minibuffer.  I do not see company-lsp in my list of enabled minor modes, either. 

**To Reproduce**
Install the latest versions of all emacs packages (including lsp-mode and company-lsp), and use the following configuration in init.el:
```
(use-package lsp-mode
  :defer t
  :config
  (require 'lsp-clients)
  (add-hook 'typescript-mode-hook 'lsp)
  (add-hook 'js2-mode-hook 'lsp))
(use-package company-lsp
  :defer t)
```

**Expected behavior**
According to the README-NEXT docs, lsp-mode should auto-initialize company-lsp if it is installed.  Completion should work upon loading a .ts or .js file, with a project root properly selected.

**Which Language Server did you use**
javascript-typescript-langserver

**OS**
Arch Linux

**Error callstack**
Info from `*Messages*` buffer pasted above.  Interestingly, since the upgrade, I now see all the JSON transactions between the langserver and emacs in `*Messages*` as well (was never there before).  I enabled `debug-on-error` and received the following error trace when trying to make my first completion:
```
Debugger entered--Lisp error: (wrong-type-argument lsp--workspace nil)
  signal(wrong-type-argument (lsp--workspace nil))
  company-lsp(prefix)
  apply(company-lsp prefix)
  company-call-backend-raw(prefix)
  apply(company-call-backend-raw prefix)
  company--force-sync(company-call-backend-raw (prefix) company-lsp)
  company-call-backend(prefix)
  company--begin-new()
  company--perform()
  company-auto-begin()
  company-idle-begin(#<buffer findings.ts> #<window 3 on findings.ts> 420 3529)
  apply(company-idle-begin (#<buffer findings.ts> #<window 3 on findings.ts> 420 3529))
  timer-event-handler([t 23561 17796 419390 nil company-idle-begin (#<buffer findings.ts> #<window 3 on findings.ts> 420 3529) nil 4000])
```
Also, I noticed that while lsp-ui is working (I LOVE the sideline, insanely helpful), the syntax highlighting is now gone from the sideline.  Should I file that under a separate issue?  Obviously not critical like this bug, but thought it was worth pointing out.

Thanks for all the hard work.  Looking forward to getting the bugs ironed out of this new version.

**Comment by cmatzenbach** [comment]
@yyoncho Ok, so I was able to get the minimal config working with a pure .js project, but I still can't get it working for typescript.  Here's the steps I took:
1. Replaced my init.el with your minimal config example
2. Cloned this .ts project - https://github.com/Microsoft/TypeScriptSamples
3. Started emacs, navigated to TypeScriptSamples/simple/animals.ts
4. Set project root as TypeScriptSamples/ folder when prompted
5. Tried `company-complete` after `this.` and also after `document.ge`, got the same error - `Company: backend company-lsp error "Wrong type argument: lsp--workspace, nil" with args (prefix)`

In my typescript project, it is also complaining about my .d.ts files in my node_modules folder, which I thought might be the issue, but I get no such warnings from this sample Microsoft project and it still does not work.  I am, however, able to complete in a normal .js project, as stated above.  Any advice, or is there a way I can provide you more information?

**Comment by yyoncho** [comment]
I tried the completion in animals.ts with the config from above and it works. The only explanation could be in your emacs doing different byte compilation.  Can you do the following: 

1. Delete ~/emacs.d/elpa
2. Start emacs and wait for dependencies to be loaded.
3. Stop emacs and delete elc files - `find ~/.emacs.d/ -name "*.elc" | xargs rm`
4. Start emacs and try to load the ts files.

**Comment by cmatzenbach** [comment]
@yyoncho That fixed it!!  Thank you so much for your help the past two days.  I really appreciate your time helping me solve these issues.  Closing now.

**Comment by yyoncho** [comment]
Reopen until the proper fix is in (deletion of lsp-mode.el) is in.

**Comment by astenman** [comment]
I have the same problem during my efforts in getting the MS Python LS work in Emacs for Windows using the new async "lsp.el". Hovering works, but when I edit the file I get

`Company: An error occurred in auto-begin
Company: backend company-lsp error "Wrong type argument: lsp--workspace, nil" with args (prefix)`

**Comment by yyoncho** [comment]
Did you try the workaround - https://github.com/emacs-lsp/lsp-mode/issues/513#issuecomment-445022923 ?

**Comment by astenman** [comment]
Yes, I just deleted the elc files for lsp-mode and company-lsp and it seems to work better ...

**Comment by cmatzenbach** [comment]
@yyoncho I have another question - in my typescript project, when I load a file, the langserver scans my node_modules folder looking at all the .d.ts files (which makes sense).  However, I just get hundreds of messages that say: `readFile /home/me/Projects/myproject/node_modules/dayjs/index.d.ts requested by TypeScript but content not available`.  It will work for awhile, but then tries to re-scan every time I start typing again, just resulting in the same thing over and over, which is killing my company completions.  Is this an issue with the front-end package, or with the langserver itself?  I'm using the latest version of everything.

**Comment by yyoncho** [comment]
I have no clue, I do not do JS/TS coding, you may file a bug in the LS project. I think that I hit the same issue with this language server, I suggest you to try https://github.com/theia-ide/typescript-language-server .

**Comment by yyoncho** [comment]
Actually this message was ignored in the previous implementation, just pushed f77194e , This will fix it.

**Comment by cmatzenbach** [comment]
@yyoncho You are the BEST.  Works perfectly.  I did notice, when you do `lsp-ui-peek-find-references`, it does this scan, finishes, times out (without pulling up the peek menu), but then if you do it again, the peek menu pulls up and you see references.  Is there a way to block this from happening on the initial call to `lsp-ui-peek-find-references`?

**Comment by yyoncho** [comment]
My guess is that the first call does not finishes in 10 second timeout and the second one is faster since the server has populated its cache. Not sure what we can do on emacs side, the alternative is switching to async methods, but still this will require some rework.

**Comment by cmatzenbach** [comment]
Ah I see.  Not a big deal really, it's still usable after you let it do it thing the first time.  Out of curiosity, what is the issue?  Is the langserver unable to parse .d.ts files (or is it an emacs issue)?  That would seem to be an essential part of any langserver for typescript - should I open a bug report there?

**Comment by yyoncho** [comment]
I am guessing here but it looks like the language server fails to parse the ts files in our timeout.

**Comment by cmatzenbach** [comment]
Ok gotcha.  I'm going to open a bug report with them.  Also, it's still being triggered I'm finding as I use it more.  It will be fine for 5 minutes or so, but when I edit the code (which forces .ts to re-scan for changes I'm sure), the whole process is repeating again.  Not sure if there's still anything you can do on your end to ignore it, or if we just have to wait for them to address the bug I'm opening.
